### PR TITLE
Port of `CallableLoci` from GATK3

### PR DESCRIPTION
**Pull request by jonn-smith**
Ported `CallableLoci` from GATK3 as an experimental tool.

I have not included tests, but have verified on a test file that the statistics and regions are basically identical to the GATK3 results.

### CODE REVIEW

**Review by lbergelson — CHANGES_REQUESTED** [review]
@jonn-smith I have some comments.  I noticed a few things that are minor bugs around the inputs, and then there are some style things I would probably change.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    /**
+     * The gap between this value and mmq are reads that are not sufficiently well mapped for calling but
+     * aren't indicative of mapping problems.  For example, if maxLowMAPQ = 1 and mmq = 20, then reads with
+     * MAPQ == 0 are poorly mapped, MAPQ >= 20 are considered as contributing to calling, where
+     * reads with MAPQ >= 1 and < 20 are not bad in and of themselves but aren't sufficiently good to contribute to
+     * calling.  In effect this reads are invisible, driving the base to the NO_ or LOW_COVERAGE states
+     */
+    @Argument(fullName = "max-low-mapq", shortName = "mlmq", 
+            doc = "Maximum value for MAPQ to be considered a problematic mapped read")
+    private byte maxLowMAPQ = 1;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:137]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    @Argument(fullName = "min-mapping-quality", shortName = "mmq", 
+            doc = "Minimum mapping quality of reads to count towards depth")
+    private byte minMappingQuality = 10;
+
+    /**
+     * Bases with less than minBaseQuality are viewed as not sufficiently high quality to contribute to the PASS state
+     */
+    @Argument(fullName = "min-base-quality", shortName = "mbq", 
+            doc = "Minimum quality of bases to count towards depth")
+    private byte minBaseQuality = 20;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:152]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+    /**
+     * If the number of QC+ bases (on reads with MAPQ > minMappingQuality and with base quality > minBaseQuality) exceeds this
+     * value and is less than maxDepth the site is considered PASS.
+     */
+    @Advanced
+    @Argument(fullName = "min-depth", shortName = "min-depth", 
+            doc = "Minimum QC+ read depth before a locus is considered callable")
+    private int minDepth = 4;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:161]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    private PrintStream outputStream = null;
+    private PrintStream summaryStream = null;
+    private Integrator integrator;
+    
+    public enum OutputFormat {
+        BED,
+        STATE_PER_BASE
+    }
+
+    public enum CalledState {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:205]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        EXCESSIVE_COVERAGE,
+        POOR_MAPPING_QUALITY
+    }
+
+    protected static class Integrator {
+        final long[] counts = new long[CalledState.values().length];
+        CallableBaseState state = null;
+    }
+
+    protected static class CallableBaseState {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:219]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+ * </p>
+ * <p/>
+ * <h3>Output</h3>
+ * <p>
+ *     A file with the callable status covering each base and a table of callable status x count of all examined bases
+ * </p>
+ * <h3>Usage example</h3>
+ * <pre>
+ *  java -jar GenomeAnalysisTK.jar \
+ *     -T CallableLoci \
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:67]: Drive by nit: this is GATK3 CLI syntax.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        POOR_MAPPING_QUALITY
+    }
+
+    protected static class Integrator {
+        final long[] counts = new long[CalledState.values().length];
+        CallableBaseState state = null;
+    }
+
+    protected static class CallableBaseState {
+        private SimpleInterval interval;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:220]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        private final CalledState state;
+
+        public CallableBaseState(SimpleInterval interval, CalledState state) {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:223]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+            this.interval = interval;
+            this.state = state;
+        }
+
+        public SimpleInterval getLocation() {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:228]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+            return interval;
+        }
+
+        public CalledState getState() {
+            return state;
+        }
+
+        public boolean changingState(CalledState newState) {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:236]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+            return state != newState;
+        }
+
+        public void update(SimpleInterval newInterval) {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:240]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+            this.interval = new SimpleInterval(
+                interval.getContig(),
+                interval.getStart(),
+                newInterval.getEnd()
+            );
+        }
+
+        @Override
+        public String toString() {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:249]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+        summary = "Collect statistics on callable, uncallable, poorly mapped, and other parts of the genome",
+        oneLineSummary = "Determine callable status of loci",
+        programGroup = CoverageAnalysisProgramGroup.class
+)
+public class CallableLoci extends LocusWalker {
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            doc = "Output file (BED or per-base format)")
+    private File outputFile = null;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:118]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        defaultFilters.add(new WellformedReadFilter());
+        defaultFilters.add(new ReadFilterLibrary.PrimaryLineReadFilter());
+        defaultFilters.add(new ReadFilterLibrary.MappedReadFilter());
+        return defaultFilters;
+    }
+
+    @Override
+    public void onTraversalStart() {
+        // Validate sample count
+        if (getHeaderForReads().getReadGroups().stream()
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:290]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+                throw new GATKException(String.format("The denominator of a ratio cannot be zero or less than zero: %d/%d", numerator, denominator));
+            }
+        }
+    }
+
+    @Override
+    public void apply(AlignmentContext alignmentContext, ReferenceContext referenceContext, FeatureContext featureContext) {
+        CalledState state;
+
+        if (BaseUtils.isNBase(referenceContext.getBase())) {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:323]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+        CallableBaseState callableState = new CallableBaseState(
+            new SimpleInterval(alignmentContext.getLocation()), 
+            state
+        );
+
+        // Update counts
+        integrator.counts[state.ordinal()]++;
+
+        if (outputFormat == OutputFormat.STATE_PER_BASE) {
+            outputStream.println(callableState.toString());
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:362]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+    /**
+     * The output of this tool will be written in this format.  The recommended option is BED.
+     */
+    @Advanced
+    @Argument(fullName = "format", shortName = "format", 
+            doc = "Output format")
+    private OutputFormat outputFormat = OutputFormat.BED;
+
+    private PrintStream outputStream = null;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:196]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+        }
+
+        // Write summary statistics
+        summaryStream.printf("%30s %s%n", "state", "nBases");
+        for (CalledState state : CalledState.values()) {
+            summaryStream.printf("%30s %d%n", state, integrator.counts[state.ordinal()]);
+        }
+
+        // Close streams
+        outputStream.close();
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:391]: 👍

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLociIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import org.broadinstitute.hellbender.testutils.ArgumentsBuilder;
+import org.broadinstitute.hellbender.testutils.IntegrationTestSpec;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.util.Arrays;
+
+public class CallableLociIntegrationTest extends CommandLineProgramTest {
+
+    public static final String testDataDir = largeFileTestDir + "/org/broadinstitute/hellbender/tools/walkers/coverage/callableloci/";
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLociIntegrationTest.java:13]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java (hunk 6 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:67]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+            outputStream = new PrintStream(outputFile);
+            summaryStream = new PrintStream(summaryFile);
+        } catch (FileNotFoundException e) {
+            throw new UserException.CouldNotCreateOutputFile("Could not create output file", e);
+        }
+
+        integrator = new Integrator();
+    }
+
+    private static double ratio(int numerator, int denominator) {
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:307]: Yeah - it was a hold-over from the old implementation.

I'll remove it.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+     * The output of this tool will be written in this format.  The recommended option is BED.
+     */
+    @Advanced
+    @Argument(fullName = "format", shortName = "format", 
+            doc = "Output format")
+    private OutputFormat outputFormat = OutputFormat.BED;
+
+    private PrintStream outputStream = null;
+    private PrintStream summaryStream = null;
+    private Integrator integrator;
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:198]: 👍

**Comment by jonn-smith** [comment]
@lbergelson @cmnbroad Thanks for the comments.  This was an interesting first experiment in using LLM-assisted coding.  Looks like it doesn't understand our codebase quite as well as I would like, though it did produce working code.  I'll see if I can tweak things in the future.

In the meantime, back to you @lbergelson for another round if you like.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 22]
```diff
@@ -46,9 +49,9 @@
  * <dt>LOW_COVERAGE</dt>
  * <dd>There were fewer than min. depth bases at the locus, after applying filters</dd>
  * <dt>EXCESSIVE_COVERAGE</dt>
- * <dd>More than -maxDepth read at the locus, indicating some sort of mapping problem</dd>
+ * <dd>More than --max-depth read at the locus, indicating some sort of mapping problem</dd>
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:52]: good catch, I didn't see these.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java [hunk 23]
```diff
@@ -338,59 +330,100 @@ public void apply(AlignmentContext alignmentContext, ReferenceContext referenceC
...
+    public void closeTool() {
+        try {
+            outputStream.close();
+            summaryStream.close();
+        } catch (IOException e) {
+            throw new GATKException("Error closing output streams", e);
+        }
+    }
+
+    private static class StateCounter {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CallableLoci.java:408]: 👍 I know it's a few more lines but I think it's much clearer.

**Review by lbergelson — APPROVED** [review]
@jonn-smith Thanks for making the changes.  I know they were kind of nitpicky but I think its better this way.

**Pull request merged**